### PR TITLE
Enhance String Diff representation in wave diffs

### PR DESCRIPTION
**Pull request by Garados007**
Before the diff window for text waves showed the full string in hexadecimal format if the entry is different in both waves.

With this change the window will only show the start of the difference and up to a maximum of 10 characters left and right. All normal letters in ASCII are shown as normal and the other letters are shown in readable notation (the name of the letter or hexadecimal form).

The window will also show which line, column and position in the full string where the first difference happen. There is no special handling of string that are only one line.

This commit also includes an example project in "internal_dev/TestTextWaveDiff".

Closing #274.

### CODE REVIEW

File: tests/Various.ipf [hunk 1]
```diff
@@ -346,6 +346,61 @@ Function TC_MMD_bck_REENTRY([md])
...
+	TC_StringDiff_Check("abc", "abcde", "0:3:3> a b c", "0:3:3> a b c d e")
+	TC_StringDiff_Check("foo\nabc\nbar", "foo\nabcde\nbar", "1:3:7> a b c <\\n>", "1:3:7> a b c d e <\\n>")
+
+	// Case 3: the line endings are different
+	TC_StringDiff_Check("abc\ndef", "abc\rdef", "0:3:3> a b c <\\n>", "0:3:3> a b c <\\r>")
+	TC_StringDiff_Check("abc\ndef", "abc\r\ndef", "0:3:3> a b c <\\n>", "0:3:3> a b c <\\r> <\\n>")
+	TC_StringDiff_Check("abc\r\ndef", "abc\rdef", "0:3:3> a b c <\\r> <\\n>", "0:3:3> a b c <\\r>")
+
+	// Case 4: one string is larger than the other one
+	TC_StringDiff_Check("text\n", "text\nabc", "1:0:5>", "1:0:5> a b c")
```
Comment (MichaelHuth) [inline tests/Various.ipf:369]: Add a test that contains an empty string and one against case-sensitivity as mentioned above.

The context trimming test is missing the cases where one side has a shorter context than MAX_STRING_DIFF_CONTEXT, e.g. no context on one side.

File: tests/Various.ipf [hunk 2]
```diff
@@ -346,6 +346,61 @@ Function TC_MMD_bck_REENTRY([md])
...
+
+	// CORE CHECKING
+
+	// Trim context
+	TC_StringDiff_Check("0123456789abcdef.0123456789abcdef", "0123456789abcdef:0123456789abcdef", "0:16:16> 6 7 8 9 a b c d e f . 0 1 2 3 4 5 6 7 8 9", "0:16:16> 6 7 8 9 a b c d e f : 0 1 2 3 4 5 6 7 8 9")
+
+	// Escaping
+	string res, expected
```
Comment (MichaelHuth) [inline tests/Various.ipf:377]: Put declarations at the beginngin of the function (or use a separate test case)

File: procedures/unit-testing-utils.ipf [hunk 3]
```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
...
 
-	return SelectString(extPrint, str, ": " + extOut)
+	// no line ending
+	marker = ""
+	return strlen(str)
 End
-#else
-static Function/S AutoExtendedStringOutput(str)
+
+static Function/S EscapeString(str)
```
Comment (MichaelHuth) [inline procedures/unit-testing-utils.ipf:1035]: Add a brief description in front of the function that it escapes purely single byte aware.

File: procedures/unit-testing-utils.ipf [hunk 4]
```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
...
+				result += "<\\n>"
+				break
+			case "\r":
+				result += "<\\r>"
+				break
+			case "\t":
+				result += "<\\t>"
+				break
+			default:
+				sprintf hex, "<%02x>", charnum
```
Comment (MichaelHuth) [inline procedures/unit-testing-utils.ipf:1067]: Would prefixing hex numbers with 0x would have an advantage?

Would using `X` instead for the format be better readable?

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
There is a bug in the code that breaks it and one location where a out-of-bounds access can happen.

Despite of these that require fixing, the overall code structure is good.

Most of the remaining comments are coding style and convention issues.

There is also typos in the commit message 
string -> strings
string -> strings.
normal letters -> displayable characters
example project -> example experiment file

File: procedures/unit-testing-utils.ipf [hunk 5]
```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
...
+
+static Function DetectEndOfLine(str, start, marker)
+	variable start
+	string str, &marker
+
+	variable i
+
+	for(i = start; i < strlen(str); i += 1)
+		if (!CmpStr(str[i], "\r"))
+			if (!CmpStr(str[i + 1], "\n"))
```
[the remark this answers is not in the document]
Comment (Garados007) [reply procedures/unit-testing-utils.ipf:1014]: This was an intended use of the Igor Feature. In other programming languages i would check the length of the string first to not read beyond its length.

I can add this check but its kind of duplicate because of the Igor behavior.

**Comment by t-b** [comment]
Am 19.10.2022 um 10:09 schrieb Max Brauer:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In procedures/unit-testing-utils.ipf 
> <https://github.com/byte-physics/igor-unit-testing-framework/pull/275#discussion_r999089314>:
> 
>> +			return i
> +		endif
> +	endfor
> +
> +	return -1
> +End
> +
> +static Function DetectEndOfLine(str, start, marker)
> +	variable start
> +	string str, &marker
> +
> +	variable i
> +
> +	for(i = start; i < strlen(str); i += 1)
> +		if (!CmpStr(str[i], "\r"))
> +			if (!CmpStr(str[i + 1], "\n"))
> 
> This was an intended use of the Igor Feature. In other programming 
> languages i would check the length of the string first to not read 
> beyond its length.
> 
> I can add this check but its kind of duplicate because of the Igor behavior.

Good reasoning. You can't know that we don't want to exploit this Igor 
behaviour. I've created 
https://github.com/byte-physics/igor-pro-coding-conventions/issues/24 to 
metion that in our coding conventions.

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
Code wise it looks good to me.

Cleanup the comments.

I think Various.pxp had no changes that require it to be included in the commit?

File: internal_dev/TestTextWaveDiff.ipf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+	wv1[2] = "Example\ntext\nwith\nunexpected\nline\nendings."
+	wv2[2] = "Example\rtext\rwith\runexpected\rline\rendings."
+
+	// the error should be on the line 3 (counting from zero)
+	wv1[3] = "Can also\r\nwork with\r\nWindows\r\nline endings."
+	wv2[3] = "Can also\r\nwork with\r\nWindows\r\nline endings!"
+
+	CHECK_EQUAL_WAVES(wv1, wv2)
+End
```
Comment (t-b) [inline internal_dev/TestTextWaveDiff.ipf:37]: Missing end-of-line at end-of-file. Please teach your editor to add that. In Igor Pro you want an empty line as last line.

**Comment by Garados007** [comment]
I have removed my changed version of `Various.pxp` from the commit.

**Review by MichaelHuth — APPROVED** [review]
